File: misc/src/main/java/Adapters/MyRelationship.java
package Adapters;

import org.neo4j.graphdb.Direction;
import org.neo4j.graphdb.Node;
import org.neo4j.graphdb.Relationship;
import org.neo4j.graphdb.RelationshipType;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

public class MyRelationship extends MyEntry implements Relationship {
    private List<RelationshipType> types = new ArrayList<>();
    private MyNode start;
    private MyNode end;

    private int minLength = 1;
    private int maxLength = 1;

    private Direction direction;

    public MyRelationship() {
    }

    public MyRelationship(MyRelationship other) {
        types = other.types;
        minLength = other.minLength;
        maxLength = other.maxLength;
        direction = other.direction;
        properties = other.properties;
    }

    public int getMinLength() {
        return minLength;
    }

    public int getMaxLength() {
        return maxLength;
    }

    public void addType(final RelationshipType type) {
        types.add(type);
    }

    public void setStart(final Node start) {
        this.start = (MyNode) start;
    }

    public void setEnd(final Node end) {
        this.end = (MyNode) end;
    }

    public void setMinLength(final int minLength) {
        this.minLength = minLength;
    }

    public void setMaxLength(final int maxLength) {
        this.maxLength = maxLength;
    }

    @Override
    public void delete() {
        throw new UnsupportedOperationException();
    }

    @Override
    public MyNode getStartNode() {
        return start;
    }

    @Override
    public MyNode getEndNode() {
        return end;
    }

    @Override
    public MyNode getOtherNode(final Node node) {
        if (node == start) {
            return end;
        }
        if (node == end) {
            return start;
        }
        return null;
    }

    @Override
    public Node[] getNodes() {
        return new Node[]{start, end};
    }

    @Override
    public RelationshipType getType() {
        throw new UnsupportedOperationException();
    }

    public Iterable<RelationshipType> getTypes() {
        return types;
    }

    @Override
    public boolean isType(final RelationshipType relationshipType) {
        return types.isEmpty() && relationshipType == null || types.contains(relationshipType);
    }

    public Direction getDirection() {
        return direction;
    }

    public void setDirection(final Direction direction) {
        this.direction = direction;
    }

    @Override
    public String toString() {
        switch (direction) {
            case INCOMING:
                return String.format("%s <-%s- %s", start, typesString(), end);
            case OUTGOING:
                return String.format("%s -%s-> %s", start, typesString(), end);
            case BOTH:
                return String.format("%s -%s- %s", start, typesString(), end);
            default:
                return super.toString();
        }
    }

    private String typesString() {
        if (this.types.isEmpty()) {
            return "";
        }

        List<String> types = new LinkedList<>();
        for (RelationshipType relationshipType : this.types) {
            types.add(relationshipType.toString());
        }

        return "[" + String.join(", ", types) + "]";
    }
}

File: src/main/kotlin/de/tubs/ips/neo4j/graph/MyRelationship.kt
package de.tubs.ips.neo4j.graph

import org.neo4j.graphdb.*

/**
 * jede relation innerhalb der Db hat eine richtung, OUTGOING oder INCOMING
 * -- und <--> haben die gleiche bedeutung...
 */

class MyRelationship(private val prototype: MyRelationshipPrototype, private val start: MyNode, private val end: MyNode, val direction: Direction) : Entity by prototype, Relationship {

    constructor(other: MyRelationship, start: MyNode, end: MyNode) : this(other.prototype, start, end, other.direction)

    init {
        start.addRelationship(this)
        end.addRelationship(this)
    }

    override fun delete() {
        throw UnsupportedOperationException()
    }

    override fun getStartNode(): MyNode {
        return start
    }

    override fun getEndNode(): MyNode {
        return end
    }

    override fun getOtherNode(node: Node): MyNode? {
        return when {
            node === start -> end
            node === end -> start
            else -> null
        }
    }

    override fun getNodes(): Array<Node> {
        return arrayOf(start, end)
    }

    override fun getType(): RelationshipType? {
        throw UnsupportedOperationException()
    }

    override fun isType(relationshipType: RelationshipType): Boolean {
        return prototype.types.contains(relationshipType)
    }

    fun typesString() : String {
        return prototype.types.joinToString("|", prefix = "[", postfix = "]", transform = { it.name() })
    }

    override fun toString(): String {
        return when (direction) {
            Direction.INCOMING -> "$start <-${typesString()}- $end <${prototype.parsingDirection}>"
            Direction.OUTGOING -> "$start -${typesString()}-> $end <${prototype.parsingDirection}>"
            Direction.BOTH -> "$start <-${typesString()}-> $end <${prototype.parsingDirection}>"
            else -> "$start -${typesString()}- $end <${prototype.parsingDirection}>"
        }
    }

    fun match(other: Relationship): Boolean {
        return prototype.matchTypes(other) && prototype.matchProperties(other)
    }
}
